Add IRMA to StaPH-B/docker-builds

jwarnn opened the pull request:
I would like to contribute this docker image of IRMA. IRMA is a CDC tool that does robust assembly, variant calling, and phasing of highly variable RNA viruses. Currently IRMA is deployed with modules for influenza, ebolavirus and coronavirus. We are going to start using this tool at Utah Public Health Laboratory.

- [ x] This comment contains a description of what is in the pull request.

- [ x] Build your own docker image using a Dockerfile
  - [ x] Directory structure should be name of the tool in lower case with special characters removed with a subdirectory of the version number (i.e. spades/3.12.0/Dockerfile)
  - [ x] Includes the recommended LABELS
- [ ] (Optional) Dockerfile is built with best practices and has been approved by a linter (such as https://hadolint.github.io/hadolint/)
- [ x] Edit main README.md
- [ x] Edit Program_Licenses.md
- [x ] Create a simple container-specific README.md in the same directory as the Dockerfile (i.e. spades/3.12.0/README.md)
- [ x] Write a GitHub actions workflow
  - [x ] Should be located in .github/workflows/ and named test-<tool>.yml (i.e. .github/workflows/test-spades.yml)
  - [ x] Any files required for building are located in the same directory as the Dockerfile (i.e. spades/3.12.0/my_spades_tests.sh)
  - [x ] Have successfully run the workflow "Test <program name> image" in your forked repository

erinyoung commented on IRMA/1.0.2/Dockerfile:
```diff
@@ -0,0 +1,77 @@
+
+# RUN executes code during the build
+# Install dependencies via apt-get or yum if using a centos or fedora base
+ENV DEBIAN_FRONTEND=noninteractive
```
I know some of the dockerfiles set this as an ENV variable, but it'd be better to change it to ARG so it disappears once the container is built

kapsakcj replied:
agreed

erinyoung commented on IRMA/1.0.2/Dockerfile:
```diff
@@ -0,0 +1,77 @@
+RUN  apt-get update && apt-get -y upgrade && apt-get -y install r-base \
+  r-base-dev
+
+RUN ls
```
this 

```
ls
```

doesn't need to be here. It just lists what is in the directory.

erinyoung commented on IRMA/1.0.2/Dockerfile:
```diff
@@ -0,0 +1,77 @@
+
+RUN unzip flu-amd-202103.zip
+
+RUN rm flu-amd-202103.zip && \
```
These three RUN statements can be combined into something like

```
RUN wget https://wonder.cdc.gov/amd/flu/irma/flu-amd-202103.zip && \ 
unzip flu-amd-202103.zip && \
rm flu-amd-202103.zip && \
chmod 755 -R flu-amd && \
mkdir data
```

erinyoung reviewed (changes requested):
Looks great!

Just needs a few edits.

Are flu sequences small enough to put in this github repo?

kapsakcj commented:
Thanks for the PR, I'm sure the CDC folks and Public Health community will be happy to see IRMA make it into the staph-b/docker-builds repo!

I have a small request, and @erinyoung will likely agree with me 😆 - could you please adjust the directory name from all caps to lowercase? `/irma` instead of `/IRMA` 

It will require changing the directory name, but also will require adjusting the GH Actions workflows to reflect the new path.

Additionally, dockerhub does not allow you to use capital letters in the docker image name, so `staphb/IRMA` will not be accepted when we deploy to dockerhub. You'll have to change the GH Actions workflows to use lowercase `irma` instead there too.

All caps IRMA is fine for titles, like the name of the GH actions workflows and the README, but there are some places that will need to be lowercase

kapsakcj commented:
I'd prefer to not add more fastqs of that size to the repo, especially when we can pull from SRA in the test layers of the docker images. Helps keep the size of the repo more manageable

erinyoung commented:
Building it locally : 
```
docker build --tag amrfinderplus:test --target test testing_dockerfiles/amrfinderplus
```
build stuff that all seems to be successful

```
Running: amrfinder --plus -n /amrfinder/test_dna.fa -p /amrfinder/test_prot.fa -g /amrfinder/test_prot.gff -O Escherichia
Software directory: '/amrfinder/'
Software version: 3.10.24
Database directory: '/amrfinder/data/2022-04-04.1'
Database version: 2022-04-04.1
AMRFinder combined translated and protein and mutation search
Running blastp...
Running hmmsearch...
Running blastx...
Running blastn...
Making report...
AMRFinder took 4 seconds to complete
exit code for test: 0
```

Running it on one of my local files:
```
/data# amrfinder       --nucleotide GCF_000006925.2_ASM692v2_genomic.fna       --threads 4       --name test       --output test_amrfinder_plus.txt -O Escherichia --plus
Running: amrfinder --nucleotide GCF_000006925.2_ASM692v2_genomic.fna --threads 4 --name test --output test_amrfinder_plus.txt -O Escherichia --plus
Software directory: '/amrfinder/'
Software version: 3.10.24
Database directory: '/amrfinder/data/2022-04-04.1'
Database version: 2022-04-04.1
AMRFinder translated nucleotide and mutation search
Running tblastn...
Running blastn...
Making report...
AMRFinder took 126 seconds to complete
```
And the results look good
```
# cat test_amrfinder_plus.txt 
Name	Protein identifier	Contig id	Start	Stop	Strand	Gene symbol	Sequence name	Scope	Element type	Element subtype	Class	Subclass	Method	Target length	Reference sequence length	% Coverage of reference sequence	% Identity to reference sequence	Alignment length	Accession of closest sequence	Name of closest sequence	HMM id	HMM description
test	NA	NC_004337.2	747917	749677	-	ipaH5	T3SS effector E3 ubiquitin-protein ligase IpaH5	plus	VIRULENCE	VIRULENCE	NA	NA	BLASTX	587	587	100.00	99.66	587	EID63069.1	T3SS effector E3 ubiquitin-protein ligase IpaH5	NA	NA
test	NA	NC_004337.2	1422067	1423779	-	ipaH3	T3SS effector E3 ubiquitin-protein ligase IpaH3	plus	VIRULENCE	VIRULENCE	NA	NA	EXACTX	571	571	100.00	100.00	571	AAN42986.1	T3SS effector E3 ubiquitin-protein ligase IpaH3	NA	NA
test	NA	NC_004337.2	1917922	1919670	+	ipaH2	T3SS effector E3 ubiquitin-protein ligase IpaH2	plus	VIRULENCE	VIRULENCE	NA	NA	EXACTX	583	583	100.00	100.00	583	AAN43435.1	T3SS effector E3 ubiquitin-protein ligase IpaH2	NA	NA
test	NA	NC_004337.2	2011995	2012324	-	emrE	multidrug efflux SMR transporter EmrE	plus	STRESS	BIOCIDE	EFFLUX	EFFLUX	BLASTX	110	110	100.00	98.18	110	CAA77936.1	multidrug efflux SMR transporter EmrE	NA	NA
test	NA	NC_004337.2	2048470	2049510	+	ipaH4	T3SS effector E3 ubiquitin-protein ligase IpaH4/H7	plus	VIRULENCE	VIRULENCE	NA	NA	PARTIALX	347	547	63.44	99.71	347	ADA74402.1	T3SS effector E3 ubiquitin-protein ligase IpaH4/H7	NA	NA
test	NA	NC_004337.2	2231196	2232236	+	ipaH2	T3SS effector E3 ubiquitin-protein ligase IpaH2	plus	VIRULENCE	VIRULENCE	NA	NA	PARTIALX	347	583	59.52	92.51	347	AAZ87985.1	T3SS effector E3 ubiquitin-protein ligase IpaH2	NA	NA
test	NA	NC_004337.2	2367772	2369127	-	glpT_E448K	Escherichia fosfomycin resistant GlpT	core	AMR	POINT	FOSFOMYCIN	FOSFOMYCIN	POINTX	452	452	100.00	99.34	452	WP_000948731.1	glycerol-3-phosphate transporter GlpT	NA	NA
test	NA	NC_004337.2	2685872	2687695	+	ipaH1	T3SS effector E3 ubiquitin-protein ligase IpaH1/H6	plus	VIRULENCE	VIRULENCE	NA	NA	BLASTX	608	608	100.00	99.34	608	ABF02983.1	T3SS effector E3 ubiquitin-protein ligase IpaH1/H6	NA	NA
test	NA	NC_004337.2	3060437	3064291	+	sigA	serine protease autotransporter toxin SigA	plus	VIRULENCE	VIRULENCE	NA	NA	EXACTX	1285	1285	100.00	100.00	1285	AAN44449.1	serine protease autotransporter toxin SigA	NA	NA
test	NA	NC_004337.2	3067740	3071855	-	pic	serine protease autotransporter toxin Pic	plus	VIRULENCE	VIRULENCE	NA	NA	EXACTX	1372	1372	100.00	100.00	1372	AFS72825.1	serine protease autotransporter toxin Pic	NA	NA
test	NA	NC_004337.2	3820849	3822570	+	iucA	aerobactin synthase IucA	plus	VIRULENCE	VIRULENCE	NA	NA	BLASTX	574	576	100.00	98.44	576	EGI25746.1	aerobactin synthase IucA	NA	NA
test	NA	NC_004337.2	3822574	3823518	+	iucB	N(6)-hydroxylysine O-acetyltransferase IucB	plus	VIRULENCE	VIRULENCE	NA	NA	BLASTX	315	315	100.00	99.68	315	AAN82074.1	N(6)-hydroxylysine O-acetyltransferase IucB	NA	NA
test	NA	NC_004337.2	3823521	3825260	+	iucC	NIS family aerobactin synthetase IucC	plus	VIRULENCE	VIRULENCE	NA	NA	BLASTX	580	580	100.00	99.48	580	AAN82073.1	NIS family aerobactin synthetase IucC	NA	NA
test	NA	NC_004337.2	3825260	3826594	+	iucD	NADPH-dependent L-lysine N(6)-monooxygenase IucD	plus	VIRULENCE	VIRULENCE	NA	NA	EXACTX	445	445	100.00	100.00	445	AAD44749.1	NADPH-dependent L-lysine N(6)-monooxygenase IucD	NA	NA
test	NA	NC_004337.2	3826603	3828795	+	iutA	ferric aerobactin receptor IutA	plus	VIRULENCE	VIRULENCE	NA	NA	EXACTX	731	731	100.00	100.00	731	AAN45165.2	ferric aerobactin receptor IutA	NA	NA
test	NA	NC_004337.2	3912669	3914708	-	sat	serine protease autotransporter toxin Sat	plus	VIRULENCE	VIRULENCE	NA	NA	INTERNAL_STOP	680	1295	52.51	99.26	680	CAE55775.1	serine protease autotransporter toxin Sat	NA	NA
test	NA	NC_004337.2	3916798	3917367	-	lpfA	long polar fimbria major subunit LpfA-O113	plus	VIRULENCE	VIRULENCE	NA	NA	EXACTX	190	190	100.00	100.00	190	AAP18950.1	long polar fimbria major subunit LpfA-O113	NA	NA
test	NA	NC_004337.2	4275284	4276372	+	pmrB_Y358N	Escherichia colistin resistant PmrB	core	AMR	POINT	COLISTIN	COLISTIN	POINTX	363	363	100.00	99.17	363	WP_001300761.1	two-component system sensor histidine kinase PmrB	NA	NA
test	NA	NC_004337.2	4482200	4483330	-	blaEC	BlaEC family class C beta-lactamase	plus	AMR	AMR	BETA-LACTAM	BETA-LACTAM	BLASTX	377	377	100.00	98.14	377	AAC77110.1	BlaEC family class C beta-lactamase	NA	NA
test	NA	NC_004851.1	36452	37237	-	virF	AraC family invasion system transcriptional regulator VirF	plus	VIRULENCE	VIRULENCE	NA	NA	EXACTX	262	262	100.00	100.00	262	ABB68924.1	AraC family invasion system transcriptional regulator VirF	NA	NA
test	NA	NC_004851.1	52748	56839	-	sepA	serine protease autotransporter toxin SepA	plus	VIRULENCE	VIRULENCE	NA	NA	EXACTX	1364	1364	100.00	100.00	1364	AAL72309.1	serine protease autotransporter toxin SepA	NA	NA
test	NA	NC_004851.1	109091	110086	-	ipaD	type III secretion system needle tip complex protein IpaD	plus	VIRULENCE	VIRULENCE	NA	NA	EXACTX	332	332	100.00	100.00	332	AAL72350.1	type III secretion system needle tip complex protein IpaD	NA	NA
test	NA	NC_004851.1	181458	183092	+	ipaH9.8	T3SS effector E3 ubiquitin-protein ligase IpaH9.8	plus	VIRULENCE	VIRULENCE	NA	NA	BLASTX	545	545	100.00	99.63	545	AAC23714.1	T3SS effector E3 ubiquitin-protein ligase IpaH9.8	NA	NA
test	NA	NC_004851.1	192554	193639	+	capU	putative hexosyltransferase CapU	plus	VIRULENCE	VIRULENCE	NA	NA	BLASTX	362	362	100.00	99.72	362	ACD06066.1	putative hexosyltransferase CapU	NA	NA
```

jwarnn commented:
I know @erinyoung is aware of this but IRMA currently will not run on paired end reads even though the documentation says it can. Maybe we should wait to merge this pull request till they have fixed this issue and if any updates are needed to the dockerfile they can be addressed. Current behavior of the software is to complete successfully on paired end reads but produce empty files.

kapsakcj commented:
Huh. That's odd about it not working on paired end files. If you want to wait, you can switch the PR to a draft state.

Have you reached out to the developers? I think Sam Sheppard?

kapsakcj commented on irma/1.0.2/Dockerfile:
```diff
@@ -0,0 +1,73 @@
+FROM ubuntu:focal as app
+
+# ARG sets environment variables during the build stage
+# USHER ver is in the conda version style.
```
please delete the comment on line 6

kapsakcj commented on irma/1.0.2/Dockerfile:
```diff
@@ -0,0 +1,73 @@
+ libxml2-dev \
+ libncurses5 \
+ wget \
+ git
```
You could clean up some `apt` garbage by adding this to line 37 & 38. Might save a little bit of space
```
git && \
apt-get autoclean && rm -rf /var/lib/apt/lists/*
```

jwarnn commented:
Tom Iverson, @tives82, has been in conversation with IRMA. They said their Illumina paired end reads work fine when they run it, most likely the problem was due to header issues in SRA fastqs.

jwarnn commented:
I have figured out a way to fix the issue from using samples from SRR. A few awk arguments need to be ran to change the headers on the fastq files. This allows IRMA to match paired end reads from Illumina. I updated the README.md with a warning of this issue with example code as a solution. I also updated the dockerfile so that the test run completes successfully with useful data. In my opinion this tool should now be ready for use!
